linalg/hypre.cpp

```diff
@@ -2969,22 +2969,34 @@ HypreParMatrix * HypreParMatrixFromBlocks(Array2D<HypreParMatrix*> &blocks,
       }
    }
 
+   MFEM_VERIFY(HYPRE_AssumedPartitionCheck(),
+               "only 'assumed partition' mode is supported");
+
    std::vector<HYPRE_BigInt> rowStarts2(2);
    rowStarts2[0] = first_loc_row;
    rowStarts2[1] = first_loc_row + all_num_loc_rows[rank];
 
-   std::vector<HYPRE_BigInt> colStarts2(2);
-   colStarts2[0] = first_loc_col;
-   colStarts2[1] = first_loc_col + all_num_loc_cols[rank];
-
-   MFEM_VERIFY(HYPRE_AssumedPartitionCheck(),
-               "only 'assumed partition' mode is supported");
+   if (first_loc_row == first_loc_col &&
+       all_num_loc_rows[rank] == all_num_loc_cols[rank])
```

Comment:
It seems very unlikely, but possible, that a rectangular matrix could have `all_num_loc_rows[rank] == all_num_loc_cols[rank]` for a particular rank but not for other ranks. Would it be better to loop over all ranks and check this condition, to set a flag `bool square`?

Reply:
That's right, I guess my point was just that even if the global matrix isn't square, doing the diagonal reordering locally doesn't seem like it would cause any problem and avoids a global communication to check that every rank is square.

Reply:
That may be a valid point about the local diagonal reordering, but another issue is that passing the same pointer for `rows` and `cols` to the 9-argument constructor will change the behavior of the constructor. These arrays are already of size equal to the number of ranks, and the communication is already done by `GatherBlockOffsetData`, so the cost to check all ranks is negligible. More important is to prevent going into the square case of the constructor on some ranks and the rectangular case on other ranks, which seems sure to cause problems.

Reply:
Yeah, that sounds fair enough. I’ll add the check in ASAP.